conda/core/index.py

```diff
@@ -55,31 +55,46 @@ def __init__(
         channels=(),
         prepend=True,
         platform=None,
+        subdirs=None,
         use_local=False,
         use_cache=False,
         unknown=None,
         prefix=None,
         repodata_fn=context.repodata_fns[-1],
+        add_system=False,
     ) -> None:
         if use_local:
             channels = ["local"] + list(channels)
         if prepend:
             channels += context.channels
-        subdirs = (platform, "noarch") if platform is not None else context.subdirs
+        if subdirs:
+            if platform:
+                log.warn("subdirs is %s, ignoring platform %s", subdirs, platform)
+        else:
+            subdirs = (platform, "noarch") if platform is not None else context.subdirs
         self.channels = {}
-        self.expanded_channels = set()
+        self.expanded_channels = []
         for channel in channels:
             urls = Channel(channel).urls(True, subdirs)
             check_allowlist(urls)
-            expanded_channels = {Channel(url) for url in urls}
+            expanded_channels = [Channel(url) for url in urls]
             self.channels[channel] = [
                 SubdirData(expanded_channel, repodata_fn=repodata_fn)
                 for expanded_channel in expanded_channels
             ]
-            self.expanded_channels |= expanded_channels
+            self.expanded_channels.extend(expanded_channels)
         self.prefix = PrefixData(prefix) if prefix else None
         self.unknown = True if unknown is None and context.offline else unknown
         self.track_features = context.track_features
+        self.add_system = add_system
+        self.system_packages = {
+            (
+                rec := _make_virtual_package(
+                    f"__{package.name}", package.version, package.build
+                )
+            ): rec
+            for package in context.plugin_manager.get_virtual_packages()
+        }
```

Comment:
I think this could be a property or a method. Doing this in initialization regardless the value of `add_system` might be expensive (e.g. the `__cuda` package has a non negligible overhead).